tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,272 @@
+                    CacheAppliedConfigurations(product, configurations);
+                }
+                catch (Exception e)
+                {
+                    Log.Warning($"Failed to apply remote configurations {e.Message}");
+                }
+            }
+
+            UnapplyRemovedConfigurations();
+
+            IEnumerable<RemoteConfiguration> GetChangedConfigurations()
+            {
+                var signed = response.Targets.Signed.Targets;
+                var targetFiles = (response.TargetFiles ?? Enumerable.Empty<RcmFile>()).ToDictionary(f => f.Path, f => f);
+
+                foreach (var kp in actualConfigPath)
+                {
+                    if (!signed.TryGetValue(kp.Key, out var signedTarget))
+                    {
+                        ThrowHelper.ThrowException($"Missing config {kp.Key} in targets");
+                    }
+
+                    if (!_products.TryGetValue(kp.Value.Product, out var product))
+                    {
+                        ThrowHelper.ThrowException($"Received config {kp.Key} for a product that was not requested");
+                    }
+
+                    var isConfigApplied = product.AppliedConfigurations.TryGetValue(kp.Key, out var appliedConfig) && appliedConfig.Hashes.SequenceEqual(signedTarget.Hashes);
+                    if (isConfigApplied)
+                    {
+                        continue;
+                    }
+
+                    yield return new RemoteConfiguration(kp.Value, targetFiles[kp.Key].Raw, signedTarget.Length, signedTarget.Hashes, signedTarget.Custom.V);
+                }
+            }
+
+            void CacheAppliedConfigurations(Product product, List<RemoteConfiguration> configurations)
+            {
+                foreach (var config in configurations)
+                {
+                    var remoteConfigurationCache = new RemoteConfigurationCache(config.Path, config.Length, config.Hashes, config.Version);
+
+                    if (product.AppliedConfigurations.ContainsKey(config.Path.Path))
+                    {
+                        product.AppliedConfigurations[config.Path.Path] = remoteConfigurationCache;
+                    }
+                    else
+                    {
+                        product.AppliedConfigurations.Add(config.Path.Path, remoteConfigurationCache);
+                    }
+                }
+            }
+
+            void UnapplyRemovedConfigurations()
+            {
+                foreach (var product in _products.Values)
+                {
+                    foreach (var appliedConfiguration in product.AppliedConfigurations)
+                    {
+                        if (!actualConfigPath.ContainsKey(appliedConfiguration.Key))
+                        {
+                            product.AppliedConfigurations.Remove(appliedConfiguration.Key);
+                        }
+                    }
+                }
```

Comment:
Removing while you iterate is dangerous in `netfx`. Need something like this I think:
```suggestion
                List<string> _toRemove = null;
                foreach (var product in _products.Values)
                {
                    foreach (var appliedConfiguration in product.AppliedConfigurations)
                    {
                        if (!actualConfigPath.ContainsKey(appliedConfiguration.Key))
                        {
                            _toRemove ??= new List<string>();
                            _toRemove.Add(appliedConfiguration.Key);
                        }
                    }
                    if(_toRemove is not null)
                    {
                        foreach(var key in _toRemove)
                        {
                            product.AppliedConfigurations.Remove(key );
                        }
                        
                        _toRemove.Clear();
                    }
                }
```

Reply:
Shame on me 😅 
Fixed wat c39a560a5c665009a67729014ec6ca05e5c7d1d3